### ISSUE TITLE
How does keras compute loss and metrics?

### ISSUE THREAD

**Issue by hardianlawi** [keras-team/keras#10455]
Im training a model with output bounded between 0 and 1 using `sigmoid` activation. Since my target is also in the interval [0,1], I use `mean_squared_error` to be my loss, and I created a customized loss function to calculate RMSE.

```
def rmse(y_true, y_pred):
    return K.sqrt(mean_squared_error(y_true, y_pred))
```

However, when I trained the model, the `rmse` seems to be less than the loss even though it s taking the root of a value between 0 and 1 which is supposedly make it higher. Any idea what could cause this?

**Comment by jadevaibhav** [comment]
Can you show your script? For now, while compiling the model, did you add loss as rmse?

**Comment by hardianlawi** [comment]
Hi @jadevaibhav ,

My script is quite long, but below is the model definition part.

```
model = create_model(**params)

model.compile(optimizer=SGD(lr=0.0001),
              loss='mean_squared_error',
              metrics=[rmse])

model.fit_generator(generator=train,
                    epochs=epochs,
                    steps_per_epoch=steps_per_epoch,
                    validation_data=val,
                    validation_steps=validation_steps,
                    workers=0,
                    max_queue_size=max_queue_size,
                    callbacks=callbacks)

save_model(model, name='model.h5', model_dir=model_dir)
```

I use `mean_squared_error` as the main loss function.

List of callbacks (I doubt these would have any effects on my loss):

```
def get_keras_callbacks(model_dir, log_dir, val_gen, nb_steps=1000, histogram_freq=0,
                        batch_size=32, write_grads=False, write_images=False, embeddings_freq=0, embeddings_layer_names=None, embeddings_metadata=None,
                        embeddings_data=None):

    model_name = 'model_{epoch:03d}_{val_rmse:.3f}.h5'

    filepath = os.path.join(model_dir, model_name)

    checkpoint = ModelCheckpoint(filepath=filepath,
                                 monitor='val_rmse',
                                 verbose=1,
                                 save_best_only=True,
                                 save_weights_only=True,
                                 mode='min')

    lr_reducer = ReduceLROnPlateau(monitor="val_rmse",
                                   factor=np.sqrt(0.5),
                                   cooldown=0,
                                   patience=1,
                                   min_lr=0.5e-6,
                                   mode='min')

    csv_logger = CSVLogger(os.path.join(log_dir, "training_log.csv"))

    early_stopping = EarlyStopping(monitor='val_rmse',
                                   min_delta=0,
                                   patience=5,
                                   verbose=0,
                                   mode='min')

    lr_scheduler = LearningRateScheduler(scheduler)

    lr_tracker = LearningRateTracker()

    # tensorboard = TensorBoard(log_dir,
    #                           histogram_freq=histogram_freq,
    #                           batch_size=batch_size,
    #                           write_grads=write_grads,
    #                           write_images=write_images,
    #                           embeddings_freq=embeddings_freq,
    #                           embeddings_layer_names=embeddings_layer_names,
    #                           embeddings_metadata=embeddings_metadata,
    #                           embeddings_data=embeddings_data)

    tensorboard = TensorBoardWrapper(batch_gen=val_gen,
                                     nb_steps=nb_steps,
                                     log_dir=log_dir,
                                     histogram_freq=histogram_freq,
                                     batch_size=batch_size,
                                     write_grads=write_grads,
                                     write_images=write_images,
                                     embeddings_freq=embeddings_freq,
                                     embeddings_layer_names=embeddings_layer_names,
                                     embeddings_metadata=embeddings_metadata,
                                     embeddings_data=embeddings_data)

    callbacks = [checkpoint, lr_reducer, csv_logger, early_stopping, lr_scheduler, lr_tracker, tensorboard]

    return callbacks
```

**Comment by jadevaibhav** [comment]
You are right, I don't see why it should be lesser unless the calculation is done on the batch, in which case it should definitely be lesser as the cumulative cost would be more than one. Try batch size of 1.

**Comment by hardianlawi** [comment]
Below are the logs of the training with batch size of 32.

```
epoch,loss,lr,rmse,val_loss,val_rmse
0,0.17613609271347522,1e-04,0.17040786737203598,0.17063588341474534,0.14131735576242208
1,0.17088977173864842,1e-04,0.16265125641208142,0.16907686630785465,0.14077091396152974
2,0.16987350876890123,1e-04,0.16200376382283865,0.16784883527755737,0.1427175907075405
3,0.16897630071528255,1e-04,0.1609579848624766,0.16647205330729484,0.13998188743442297
4,0.16842497369125486,1e-04,0.16085571061708034,0.16590696757733822,0.14194387470185757
5,0.16775066115036608,1e-04,0.16009323768876493,0.16592117716372012,0.14159304676800966
6,0.16736669797450304,1e-04,0.16015340986736118,0.16450290610790252,0.1419200500383973
7,0.16683438361026345,1e-04,0.15954565066546203,0.16476466418504715,0.1426806757763028
8,0.16659767069146036,1e-04,0.1597313535053283,0.16393736211061477,0.14259959346950055
```

I'm not sure if I understand what you mean by `the calculation is done on the batch`, but I believe `mean_squared_error` will always be less than 1 since the target is in the interval [0,1], thus taking the root should result in a higher value.

**Comment by jadevaibhav** [comment]
What I meant was in batch-wise computations, the loss is accumulated and updated at the end of the batch. I think metric value(here rmse) is calculated in the same manner, which would be normalized by batch size(dividing by batch size). As cumulative error would be greater than 1, it would result in lesser rmse. But, after normalizing, the values would be less than 1, which is displayed. Anyway, this is my guess at the implementation, but, I think it logically fits. Hope that answered it.

**Comment by jadevaibhav** [comment]
Please tell me if the issue is resolved. Just try the batch size of 1 and please post the results if the rmse is still lesser than the mse.

**Comment by hardianlawi** [comment]
```
What I meant was in batch-wise computations, the loss is accumulated and updated at the end of the batch. I think metric value(here rmse) is calculated in the same manner, which would be normalized by batch size(dividing by batch size)
```

@jadevaibhav regarding your argument about the batch-wise computations, from my understanding, that's just how batch gradient descent works, i.e. for every step (batch of size n) you compute the loss for that step and update your weights accordingly. In addition to that, as you say, it is divided by the batch size since it is called `mean_squared_error`.

```
As cumulative error would be greater than 1, it would result in lesser rmse. But, after normalizing, the values would be less than 1, which is displayed.
```

Though the cumulative error before dividing it is greater than 1, the root in `rmse` is performed after dividing the cumulative error by the batch size which is why your reasoning does not seem to fit here.

I would very much like to try training with `batch_size` of 1, but it appears that training with just one sample every step is too noise for my model and thus I often end up with `nan`. However, when I observe the values during the training, still, the `rmse` doesn't reflect the way I think of how the metrics is calculated, i.e. using the same data for the loss, to calculate the metrics which should be just the `sqrt` of the loss. Even if the metrics is calculated after the weights update, it still does not make sense to me that  the `rmse` could be less than the loss.
